### PR TITLE
CLDC-1688 scheme change mailer

### PR DESCRIPTION
**Pull request by CheshireSwift**
Extracted just the E-mail parts of #1029, updated URL to match #1035.

As per #1029, E-mail template exists and is configured to match designs 🙂

### CODE REVIEW

File: app/controllers/locations_controller.rb [hunk 1]
```diff
@@ -48,8 +48,16 @@ def deactivate_confirm
   end
 
   def deactivate
-    if @location.location_deactivation_periods.create!(deactivation_date: params[:deactivation_date]) && reset_location_and_scheme_for_logs!
+    if @location.location_deactivation_periods.create!(deactivation_date: params[:deactivation_date])
```
Comment (CheshireSwift) [inline app/controllers/locations_controller.rb:51]: This technically removes `reset_location_...` from the conditional, but since it's using `update!` it already either returns true or throws, so behaviour should be identical.

File: app/controllers/schemes_controller.rb [hunk 2]
```diff
@@ -49,8 +49,15 @@ def deactivate_confirm
   end
 
   def deactivate
-    if @scheme.scheme_deactivation_periods.create!(deactivation_date: params[:deactivation_date]) && reset_location_and_scheme_for_logs!
+    if @scheme.scheme_deactivation_periods.create!(deactivation_date: params[:deactivation_date])
```
Comment (CheshireSwift) [inline app/controllers/schemes_controller.rb:52]: (Same comment as `locations_controller` re: removing `reset_location_and_scheme_for_logs!` from condition.)

File: app/mailers/location_or_scheme_deactivation_mailer.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+class LocationOrSchemeDeactivationMailer < NotifyMailer
+  DEACTIVATION_TEMPLATE_ID = "8d07a8c3-a4e3-4102-8be7-4ee79e4183fd".freeze
+
+  def send_deactivation_mail(user, log_count, update_logs_url, scheme_name, postcode = nil)
+    send_email(
+      user.email,
+      DEACTIVATION_TEMPLATE_ID,
+      {
+        log_count:,
+        log_or_logs: log_count == 1 ? "log" : "logs",
```
Comment (jamesrose) [inline app/mailers/location_or_scheme_deactivation_mailer.rb:10]: You can use https://apidock.com/rails/ActionView/Helpers/TextHelper/pluralize directly in the template here :)

Comment (CheshireSwift) [reply app/mailers/location_or_scheme_deactivation_mailer.rb:10]: When you say "in the template"; this is going out to Notify rather than ERB internally. Do we have access to it actually in the Notify template (as configured in the web interface), or do you just mean we can put `log_or_logs: pluralize(log_count, "log")` here?

tl;dr do we have access to these sort of helpers in Notify templates?

File: app/mailers/location_or_scheme_deactivation_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    counts_by_user(logs).each do |user, count|
+      send_deactivation_mail(user, count, update_logs_url, scheme_name, postcode) if user
+    end
+  end
+
+private
+
+  def counts_by_user(logs)
+    logs.each_with_object(Hash.new(0)) do |log, counts|
+      counts[log.created_by] += 1
```
Comment (kosiakkatrina) [inline app/mailers/location_or_scheme_deactivation_mailer.rb]: Is this going to create a hash with a key as an entire user or am I missing something?

**Review by kosiakkatrina — APPROVED** [review]
Looks good!

File: app/controllers/locations_controller.rb [hunk 5]
```diff
@@ -48,8 +48,16 @@ def deactivate_confirm
...
+      logs = reset_location_and_scheme_for_logs!
+
       flash[:notice] = deactivate_success_notice
+      LocationOrSchemeDeactivationMailer.new.send_deactivation_mails(
+        logs.to_a,
```
Comment (jamesrose) [inline app/controllers/locations_controller.rb]:
```suggestion
        logs,
```

Why `to_a`?

File: app/controllers/schemes_controller.rb [hunk 6]
```diff
@@ -49,8 +49,15 @@ def deactivate_confirm
...
+      logs = reset_location_and_scheme_for_logs!
+
       flash[:notice] = deactivate_success_notice
+      LocationOrSchemeDeactivationMailer.new.send_deactivation_mails(
+        logs.to_a,
+        url_for(controller: "lettings_logs", action: "update_logs"),
+        @scheme.service_name,
+      )
```
Comment (jamesrose) [inline app/controllers/schemes_controller.rb]:
```suggestion
      LocationOrSchemeDeactivationMailer.new.send_deactivation_mails(
        logs,
        url_for(controller: "lettings_logs", action: "update_logs"),
        @scheme.service_name,
      )
```

File: app/mailers/location_or_scheme_deactivation_mailer.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    end
+  end
```
Comment (jamesrose) [inline app/mailers/location_or_scheme_deactivation_mailer.rb]:
```suggestion
  def send_deactivation_mails(logs, update_logs_url, scheme_name, postcode = nil)
        logs.group(:created_by).count.compact.each do |user, count|
      send_deactivation_mail(user, count, update_logs_url, scheme_name, postcode)
    end
  end

private
```

You can do the whole thing in ActiveRecord :)

File: config/routes.rb [hunk 8]
```diff
@@ -111,6 +111,7 @@
       get "csv-download", to: "lettings_logs#download_csv"
       post "email-csv", to: "lettings_logs#email_csv"
       get "csv-confirmation", to: "lettings_logs#csv_confirmation"
+      get "update-logs", to: "lettings_logs#update_logs"
```
Comment (jamesrose) [inline config/routes.rb]: Does this route anywhere? I don't see the corresponding action.

File: app/mailers/location_or_scheme_deactivation_mailer.rb (hunk 4 above)
Comment (CheshireSwift) [reply app/mailers/location_or_scheme_deactivation_mailer.rb]: It is, but James has highlighted an alternate approach for this anyway :)

File: config/routes.rb (hunk 8 above)
Comment (CheshireSwift) [reply config/routes.rb]: It doesn't exist yet; I pulled this from Kat's PR to actually add the page. Without at least adding it to the routing, the URL construction for the E-mail fails. This seemed like the tidiest place to draw the line.

File: app/controllers/locations_controller.rb (hunk 5 above)
Comment (CheshireSwift) [reply app/controllers/locations_controller.rb]: I think at one point during development it made the ergonomics of the tests easier, but that no longer appears to be the case. Happy to remove.

**Review by jamesrose — APPROVED** [review]
